### PR TITLE
IA-1757 Back Leo: add disk PubSub messaging and algebra

### PR DESCRIPTION
**Pull request by gpcarr**
PR to add message handler methods to LeoPubsubSubscriber for `CreateDisk`, `DeleteDisk` and `UpdateDisk` 

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: remove `case _` and add the `DeleteDiskMessage` and `UpdateDiskMessage`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: if `newSize` is None, is there still point of having this message?

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/MockGoogleDiskService.scala
Comment (Qi77Qi) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/MockGoogleDiskService.scala:12]: we should probably have added this in wb-libs similar to https://github.com/broadinstitute/workbench-libs/blob/develop/google2/src/test/scala/org/broadinstitute/dsde/workbench/google2/mock/FakeGoogleStorageInterpreter.scala

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala
Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: I guess not.. should we make `newSize` not be an Option?

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/MockGoogleDiskService.scala
Comment (gpcarr) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/MockGoogleDiskService.scala:12]: I can add a `TODO` similar to https://github.com/DataBiosphere/leonardo/blob/develop/http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/MockGoogleComputeService.scala#L24 and we can create a tech debt ticket to move them both to wb-libs

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (gpcarr) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: @akarukappadath @rtitle do I need to check disk status before deleting? and what about for updating disk? what statuses are we checking for?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: I don't think you need to set an id in the create request. This field is output only, meaning Google will generate its own ID.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: Same here, I think `creationTimestamp` is output only

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: Probably want to use `.asString` or `.value` for these case classes. `.toString` may not generate the format we are expecting.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: I'm picturing a flow like the following here (psuedocode):
```
val createDisk = for {
  googleResp <- googleService.createDisk(…)
  _ <- // update persistent disk in the DB with async fields like `googleId`
  _ <- // monitor disk creation
} yield ()

createDisk.handleErrorWith { 
  // if an error occurred, log and set the disk status to `Error`
}
```

See `handleCreateRuntimeMessage` as a model. I think monitoring stuff is tracked as a separate ticket (and will change once https://github.com/DataBiosphere/leonardo/pull/1395/ is merged).

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: I'm also wondering if we want to make Google calls inline here in the pubsub subscriber, or create a separate algebra/interpreter similar to `RuntimeAlgebra` and `Gce/DataprocInterpreter`. I'm not sure I see the need for a separate algebra yet, so maybe let's keep things simple for now and just make the calls here in `LeoPubsubMessageSubscriber`.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: I think we expect the disk to be in `Deleting` state by the time the subscriber receives the `delete` message. So maybe check the record for `Deleting` status, and raise an error otherwise (this is similar to `handleDeleteRuntimeMessage`).

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:370]: Just FYI, another way to write this more concisely is:
```
F.fromEither(diskOpt.toRight(PubsubHandleMessageError.DiskNotFound(msg.diskId, msg)))
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala
Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: yea...make it not optional so that if we're sending this event, it's an event that we do care.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: Maybe 
```
_ <- msg.newSize.traverse_ { newSize =>
  googleDiskService.resizeDisk(disk.googleProject, disk.zone, disk.name, newSize.gb)
}
```

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:418]: Maybe we make a `DiskInvalidState` exception type for the status exceptions above

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: That's a good point, maybe we should make it non-optional.
If there are multiple potential fields to update, then we would make them optional.

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriberSpec.scala
Comment (rtitle) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriberSpec.scala:22]: Probably want to add some test cases in this spec for the new `handle` methods in `LeoPubsubMessageSubscriber`

**Comment by rtitle** [comment]
Left some suggestions above, this is a good first pass though. It really feels like we're getting closer with the PD work.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: good to know, will update!

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala
Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: should I also make it non-optional in `UpdateDiskRequest`? https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala#L305 (I removed `diskType` and `blockSize` from here as well)

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: it needs to be optional in request because it's possible that they just want to update `diskType` or `blockSize`

Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: Oh so I shouldn't remove `diskType` or `blockSize` from the request? the reason i removed them is because Google doesn't support updating `diskType` or `blockSize`

Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: would we want to allow users to update `diskType` snd `blockSize` just in our DB?

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/leoPubsubMessageSubscriberModels.scala]: hmmm if that's the case..yea probably should just remove them from the request and make size non optional...cc @rtitle for confirmation

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala]: we should just call this `size`...being an `UpdateDiskRequest` already indicates this is for updating the disk size. terra UI preferred the field matches other requests so that they don't have to adjust the field name differently depending on the call they're making

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:342]: use `_` since doesn't seem like you care about `diskResult` here

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala
Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/DiskRoutes.scala]: will update :)

**Comment by gpcarr** [comment]
jenkins retest

**Comment by gpcarr** [comment]
jenkins retest

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/PersistentDiskComponent.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/PersistentDiskComponent.scala:115]: just curious, when does `googleId` change?

**Review by Qi77Qi — APPROVED** [review]
lgtm 👍

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/PersistentDiskComponent.scala
Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/PersistentDiskComponent.scala:115]: We create the Disk in our db, then send the api call to google which creates the `googleId` so we have to update the `googleId` in our db after we successfully create the disk in google

**Comment by gpcarr** [comment]
@Qi77Qi @rtitle looks like the jenkins tests failed because the `fiab-start` job failed but the `leo` build was successful (https://fc-jenkins.dsp-techops.broadinstitute.org/job/swatomation-pipeline/34686/). seems like I should wait until this job is fixed before merging according to a thread in #dsp-devops-champions

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: this is no longer an optional field so no need to check anymore if it exists :)

**Review by akarukappadath — APPROVED** [review]
lgtm!

**Comment by gpcarr** [comment]
jenkins retest

**Comment by gpcarr** [comment]
jenkins retest

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:346]: For `createDisk`, don't we have `blockSize`, `diskType` fields too?

**Review by rtitle — APPROVED** [review]
Looks great!

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:346]: oh we do, the corresponding fields in google would be `type` and `physicalBlockSizeBytes` I believe. I need to add this to wb-libs as well (it'll be a different `diskClient.insertDisk` method we need to use), should I create a new ticket to add this to wb-libs and then add `blockSize` and `diskType` in the leo create disk call? Alternatively, i can hold off on merging this PR, make the changes in wb-libs and then add the leo changes to this PR.

Comment (gpcarr) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:346]: Oh wait, nevermind I think I can just make the changes in leo since we pass a `Disk` object to the google API call

Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala:346]: Yup, latest change looks good 👍

**Pull request merged**
